File: system-test/src/testkit/dsl/core/usecase/usecases/BrowseCoupons.ts
import type { BookShopDriver } from '../../../../driver/port/book-shop-driver.js';
import type { BrowseCouponsResponse } from '../../../../driver/port/dtos/BrowseCouponsResponse.js';
import { UseCaseResult } from '../../shared/use-case-result.js';
import type { UseCaseContext } from '../../shared/use-case-context.js';
import { BaseBookShopUseCase } from './base/BaseBookShopUseCase.js';
import { BrowseCouponsVerification } from './BrowseCouponsVerification.js';

export class BrowseCoupons extends BaseBookShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification> {
  constructor(driver: BookShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  async execute(): Promise<UseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification>> {
    const result = await this.driver.browseCoupons({});

    return new UseCaseResult(
      result,
      this.context,
      (response, ctx) => new BrowseCouponsVerification(response, ctx),
    );
  }
}

File: system-test/src/main/java/com/jcupac/bookshop/testkit/dsl/core/usecase/usecases/BrowseCoupons.java
package com.jcupac.bookshop.testkit.dsl.core.usecase.usecases;

import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseContext;
import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseResult;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.base.BaseBookShopUseCase;
import com.jcupac.bookshop.testkit.driver.port.BookShopDriver;
import com.jcupac.bookshop.testkit.driver.port.dtos.BrowseCouponsRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.BrowseCouponsResponse;

public class BrowseCoupons extends BaseBookShopUseCase<BrowseCouponsResponse, BrowseCouponsVerification> {
    public BrowseCoupons(BookShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    @Override
    public UseCaseResult<BrowseCouponsResponse, BrowseCouponsVerification> execute() {
        var request = BrowseCouponsRequest.builder().build();
        var result = driver.browseCoupons(request);
        return new UseCaseResult<>(result, context, BrowseCouponsVerification::new);
    }
}
